### PR TITLE
fix: psbt bare utxo signing

### LINKED ISSUE #3897: Failed to sign PSBT transaction with Bare MultiSig UTXO
Our platform is currently developing a SRC20 marketplace and plans to integrate with the Hiro wallet. During our testing, we found that Hiro cannot sign PSBT transactions that include Bare MultiSig outputs (similar to the 2nd and 3rd outputs in this transaction: https://mempool.space/tx/5b05ce32451f80ba0beb8887a0ba901c3f1c553d594e72fce7d4d50c5b557c48). The error we encountered is shown in the following screenshot. 

Could you please assist us in understanding what may be causing this issue? 
By the way, we have tested with the UniSat wallet and it is able to sign these transactions correctly. However, our first choice is still Hiro, so we hope that you can assist us in resolving this issue. Thanks.

image

### LINKED ISSUE #4215: Unexpected application error listing transferring stamp or SRC-20 on openstamp.io with Leather
Steps to reproduce
1. Sign in on openstamp.io with a wallet that has stamps/SRC-20 tokens
2. Go to assets via the top right menu
3. Transfer a token, click confirm
4. see error (note: if the wallet does not have enough BTC to cover the fee no error will be displayed accept in the console)

Issue with a Stamp:
image

`Unexpected Application Error!
Unknown address type=unknown
Error: Unknown address type=unknown
    at Object.encode (chrome-extension://ldinpeekobnhjjdofggfgjlcehhmanlj/579.js:200337:19)
    at getAddressFromOutScript (chrome-extension://ldinpeekobnhjjdofggfgjlcehhmanlj/index.js:47508:93)
    at chrome-extension://ldinpeekobnhjjdofggfgjlcehhmanlj/index.js:23824:76
    at Array.map (<anonymous>)
    at chrome-extension://ldinpeekobnhjjdofggfgjlcehhmanlj/index.js:23819:42
    at Object.useMemo (chrome-extension://ldinpeekobnhjjdofggfgjlcehhmanlj/579.js:144338:189)
    at exports.useMemo (chrome-extension://ldinpeekobnhjjdofggfgjlcehhmanlj/579.js:150572:208)
    at useParsedOutputs (chrome-extension://ldinpeekobnhjjdofggfgjlcehhmanlj/index.js:23819:27)
    at useParsedPsbt (chrome-extension://ldinpeekobnhjjdofggfgjlcehhmanlj/index.js:23958:25)
    at PsbtSigner (chrome-extension://ldinpeekobnhjjdofggfgjlcehhmanlj/index.js:24093:7)
    at Xh (chrome-extension://ldinpeekobnhjjdofggfgjlcehhmanlj/579.js:144321:137)
    at Wk (chrome-extension://ldinpeekobnhjjdofggfgjlcehhmanlj/579.js:144444:337)
    at Vk (chrome-extension://ldinpeekobnhjjdofggfgjlcehhmanlj/579.js:144434:389)
    at Uk (chrome-extension://ldinpeekobnhjjdofggfgjlcehhmanlj/579.js:144434:320)
    at Jk (chrome-extension://ldinpeekobnhjjdofggfgjlcehhmanlj/579.js:144434:180)
    at Ok (chrome-extension://ldinpeekobnhjjdofggfgjlcehhmanlj/579.js:144425:88)
    at Fk (chrome-extension://ldinpeekobnhjjdofggfgjlcehhmanlj/579.js:144427:300)
    at jg (chrome-extension://ldinpeekobnhjjdofggfgjlcehhmanlj/579.js:144281:105)
    at Xk (chrome-extension://ldinpeekobnhjjdofggfgjlcehhmanlj/579.js:144437:470)
    at Qk (chrome-extension://ldinpeekobnhjjdofggfgjlcehhmanlj/579.js:144435:398)
    at Hk (chrome-extension://ldinpeekobnhjjdofggfgjlcehhmanlj/579.js:144424:269)
    at J (chrome-extension://ldinpeekobnhjjdofggfgjlcehhmanlj/579.js:158098:203)
    at MessagePort.R (chrome-extension://ldinpeekobnhjjdofggfgjlcehhmanlj/579.js:158099:128)`

Issue as seen with an SRC-20 transfer
image

### PR DESCRIPTION
**Pull request by fbwoolf**
> Try out this version of Leather — [download extension builds](https://github.com/leather-wallet/extension/actions/runs/6937852174).

Solution to handle bare UTXO signing.

The app has to initialize the tx with `disableScriptCheck: true`, and the changes here would handle getting the address from script type 'ms' w/out errors.

Todo:
- [x] Put up issue in btc-signer lib to handle all types there.

EDIT: Issue opened here: https://github.com/paulmillr/scure-btc-signer/issues/55

### CODE REVIEW

File: test-app/src/components/bitcoin.tsx [hunk 1]
```diff
@@ -43,10 +43,13 @@ function getTaprootPayment(publicKey: Uint8Array) {
   return btc.p2tr(ecdsaPublicKeyToSchnorr(publicKey), undefined, bitcoinTestnet);
 }
 
+let bareMultiSigLockingScript =
```
Comment (fbwoolf) [inline test-app/src/components/bitcoin.tsx]: To remove, just for testing.

File: src/shared/crypto/bitcoin/bitcoin.utils.ts [hunk 2]
```diff
@@ -64,24 +65,19 @@ export function decodeBitcoinTx(tx: string) {
 export function getAddressFromOutScript(script: Uint8Array, bitcoinNetwork: BtcSignerNetwork) {
   const outputScript = btc.OutScript.decode(script);
```
Comment (pete-watters) [inline src/shared/crypto/bitcoin/bitcoin.utils.ts:69]: How about changing this to use a `switch` statement for readability? 🤔

Comment (fbwoolf) [reply src/shared/crypto/bitcoin/bitcoin.utils.ts:69]: Yeah, prob a good idea, will do. 👍

**Comment by fbwoolf** [comment]
@kyranjamie what do you think about this response?
https://github.com/paulmillr/scure-btc-signer/issues/55

File: src/shared/crypto/bitcoin/bitcoin.utils.ts (hunk 2 above)
Comment (pete-watters) [reply src/shared/crypto/bitcoin/bitcoin.utils.ts:69]: I think that looks much better but yesterday @kyranjamie suggested I don't use `switch` on my PR in favour of early return. 

I think this is a nice situation for `switch` but maybe we should keep consistency with early return. 

Sorry @fbwoolf to ask you to change it then back again 🤦

**Review by pete-watters — APPROVED** [review]

**Comment by kyranjamie** [comment]
Where does `ms` come from in the first place @fbwoolf ?

**Comment by fbwoolf** [comment]
> Where does `ms` come from in the first place @fbwoolf ?

`ms` is the outscript `type` supplied by the app trying to do multisig. Apparently, though, they are doing it wrong, see this comment on issue: https://github.com/hirosystems/wallet/issues/3897#issuecomment-1658998229

~~EDIT: It appears I need to refactor this and NOT handle `ms` this way. The lib has chosen not to handle it for a reason. Link to comment on issue with example script: https://github.com/hirosystems/wallet/issues/3897#issuecomment-1639222101~~ It does seem like this PR fixes the issue for those using bitcoinjs, so I think we should merge it?

File: src/shared/crypto/bitcoin/bitcoin.utils.ts (hunk 2 above)
Comment (fbwoolf) [reply src/shared/crypto/bitcoin/bitcoin.utils.ts:69]: @kyranjamie what is the pref here?

Comment (kyranjamie) [reply src/shared/crypto/bitcoin/bitcoin.utils.ts:69]: I have no trouble with switch statements in general. This is fine imo. In this instance we're passing in a variable primitive and using the cases for matchers (versus passing in a static value, and adding expressions to the case statements)

Pete is referring to a comment on

```ts
switch(true) {}
```

File: src/shared/crypto/bitcoin/bitcoin.utils.ts [hunk 3]
```diff
@@ -64,35 +65,33 @@ export function decodeBitcoinTx(tx: string) {
...
 
-  if (outputScript.type === 'pk' || outputScript.type === 'tr') {
-    return btc.Address(bitcoinNetwork).encode({
-      type: outputScript.type,
-      pubkey: outputScript.pubkey,
-    });
+  switch (outputScript.type) {
+    case 'pkh' || 'sh' || 'wpkh' || 'wsh':
```
Comment (fbwoolf) [inline src/shared/crypto/bitcoin/bitcoin.utils.ts]: This will likely get closed, but reminder this is written incorrectly and needs to be fixed.

**Review by pete-watters — APPROVED** [review]

**Pull request merged**